// flags/dist/next.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }




var _chunkWECCP5XGcjs = require('./chunk-WECCP5XG.cjs');










var _chunkZFCSUOEScjs = require('./chunk-ZFCSUOES.cjs');
require('./chunk-MEG7RDX7.cjs');

// src/next/index.ts
var _cookies = require('@edge-runtime/cookies');

// src/next/is-internal-next-error.ts
var REACT_POSTPONE_TYPE = /* @__PURE__ */ Symbol.for("react.postpone");
function isPostpone(error) {
  return typeof error === "object" && error !== null && "$$typeof" in error && error.$$typeof === REACT_POSTPONE_TYPE;
}
function isInternalNextError(error) {
  if (isPostpone(error)) return true;
  if (typeof error !== "object" || error === null || !("digest" in error) || typeof error.digest !== "string") {
    return false;
  }
  const errorCode = error.digest.split(";")[0];
  return errorCode === "NEXT_REDIRECT" || errorCode === "DYNAMIC_SERVER_USAGE" || errorCode === "BAILOUT_TO_CLIENT_SIDE_RENDERING" || errorCode === "NEXT_NOT_FOUND";
}

// src/next/overrides.ts
var memoizedDecrypt = _chunkWECCP5XGcjs.memoizeOne.call(void 0, 
  (text) => _chunkZFCSUOEScjs.decryptOverrides.call(void 0, text),
  (a, b) => a[0] === b[0],
  // only the first argument gets compared
  { cachePromiseRejection: true }
);
async function getOverrides(cookie) {
  if (typeof cookie === "string" && cookie !== "") {
    const cookieOverrides = await memoizedDecrypt(cookie);
    return _nullishCoalesce(cookieOverrides, () => ( null));
  }
  return null;
}

// src/next/precompute.ts
async function evaluate(flags) {
  return Promise.all(flags.map((flag2) => flag2()));
}
async function precompute(flags) {
  const values = await evaluate(flags);
  return serialize2(flags, values);
}
function combine(flags, values) {
  return Object.fromEntries(flags.map((flag2, i) => [flag2.key, values[i]]));
}
async function serialize2(flags, values, secret = process.env.FLAGS_SECRET) {
  if (!secret) {
    throw new Error("flags: Can not serialize due to missing secret");
  }
  if (flags.length === 0) return "__no_flags__";
  return _chunkWECCP5XGcjs.serialize.call(void 0, combine(flags, values), flags, secret);
}
async function deserialize2(flags, code, secret = process.env.FLAGS_SECRET) {
  if (!secret) {
    throw new Error("flags: Can not serialize due to missing secret");
  }
  if (code === "__no_flags__") return {};
  return _chunkWECCP5XGcjs.deserialize.call(void 0, code, flags, secret);
}
async function getPrecomputed(flagOrFlags, precomputeFlags, code, secret = process.env.FLAGS_SECRET) {
  if (!secret) {
    throw new Error(
      "flags: getPrecomputed was called without a secret. Please set FLAGS_SECRET environment variable."
    );
  }
  if (code === "__no_flags__") {
    const keys = Array.isArray(flagOrFlags) ? flagOrFlags.map((f) => f.key).join(", ") : flagOrFlags.key;
    console.warn(
      `flags: getPrecomputed was called with a code generated from an empty flags array. The flag(s) "${keys}" can not be resolved. Make sure to include them in the array passed to serialize/precompute.`
    );
  }
  const flagSet = await deserialize2(precomputeFlags, code, secret);
  if (Array.isArray(flagOrFlags)) {
    return flagOrFlags.map((flag2) => {
      if (!Object.hasOwn(flagSet, flag2.key)) {
        console.warn(
          `flags: Tried to read precomputed value for flag "${flag2.key}" which is not part of the precomputed flags. Make sure to include it in the array passed to serialize/precompute.`
        );
      }
      return flagSet[flag2.key];
    });
  } else {
    const key = flagOrFlags.key;
    if (!Object.hasOwn(flagSet, key)) {
      console.warn(
        `flags: Tried to read precomputed value for flag "${key}" which is not part of the precomputed flags. Make sure to include it in the array passed to serialize/precompute.`
      );
    }
    return flagSet[key];
  }
}
function* cartesianIterator(items) {
  const remainder = items.length > 1 ? cartesianIterator(items.slice(1)) : [[]];
  for (const r of remainder) for (const h of items.at(0)) yield [h, ...r];
}
async function generatePermutations(flags, filter = null, secret = process.env.FLAGS_SECRET) {
  if (!secret) {
    throw new Error(
      "flags: generatePermutations was called without a secret. Please set FLAGS_SECRET environment variable."
    );
  }
  if (flags.length === 0) return ["__no_flags__"];
  const options = flags.map((flag2) => {
    if (!flag2.options) return [false, true];
    return flag2.options.map((option) => option.value);
  });
  const list = [];
  for (const permutation of cartesianIterator(options)) {
    const permObject = permutation.reduce(
      (acc, value, index) => {
        acc[flags[index].key] = value;
        return acc;
      },
      {}
    );
    if (!filter || filter(permObject)) list.push(permObject);
  }
  return Promise.all(list.map((values) => _chunkWECCP5XGcjs.serialize.call(void 0, values, flags, secret)));
}

// src/next/create-flags-discovery-endpoint.ts
function createFlagsDiscoveryEndpoint(getApiData, options) {
  return async (request) => {
    const access = await _chunkZFCSUOEScjs.verifyAccess.call(void 0, 
      request.headers.get("Authorization"),
      _optionalChain([options, 'optionalAccess', _ => _.secret])
    );
    if (!access) return Response.json(null, { status: 401 });
    const apiData = await getApiData(request);
    return new Response(JSON.stringify(apiData), {
      headers: {
        "x-flags-sdk-version": _chunkZFCSUOEScjs.version,
        "content-type": "application/json"
      }
    });
  };
}

// src/next/dedupe.ts
function createCacheNode() {
  return {
    s: 0 /* UNTERMINATED */,
    v: void 0,
    o: null,
    p: null
  };
}
var cacheRegistry = /* @__PURE__ */ new WeakMap();
function dedupe(fn) {
  const requestStore = /* @__PURE__ */ new WeakMap();
  const dedupedFn = async function(...args) {
    const { headers } = await Promise.resolve().then(() => _interopRequireWildcard(require("next/headers")));
    const h = await headers();
    let cacheNode = requestStore.get(h);
    if (!cacheNode) {
      cacheNode = createCacheNode();
      requestStore.set(h, cacheNode);
    }
    for (let i = 0; i < args.length; i++) {
      const arg = args[i];
      if (typeof arg === "function" || typeof arg === "object" && arg !== null) {
        let objectCache = cacheNode.o;
        if (objectCache === null) {
          cacheNode.o = objectCache = /* @__PURE__ */ new WeakMap();
        }
        const objectNode = objectCache.get(arg);
        if (objectNode === void 0) {
          cacheNode = createCacheNode();
          objectCache.set(arg, cacheNode);
        } else {
          cacheNode = objectNode;
        }
      } else {
        let primitiveCache = cacheNode.p;
        if (primitiveCache === null) {
          cacheNode.p = primitiveCache = /* @__PURE__ */ new Map();
        }
        const primitiveNode = primitiveCache.get(arg);
        if (primitiveNode === void 0) {
          cacheNode = createCacheNode();
          primitiveCache.set(arg, cacheNode);
        } else {
          cacheNode = primitiveNode;
        }
      }
    }
    if (cacheNode.s === 1 /* TERMINATED */) {
      return cacheNode.v;
    }
    if (cacheNode.s === 2 /* ERRORED */) {
      throw cacheNode.v;
    }
    try {
      const result = fn.apply(this, args);
      cacheNode.s = 1 /* TERMINATED */;
      cacheNode.v = result;
      return result;
    } catch (error) {
      cacheNode.s = 2 /* ERRORED */;
      cacheNode.v = error;
      throw error;
    }
  };
  cacheRegistry.set(dedupedFn, requestStore);
  return dedupedFn;
}
async function clearDedupeCacheForCurrentRequest(dedupedFn) {
  if (typeof dedupedFn !== "function") {
    throw new Error("dedupe: not a function");
  }
  const requestStore = cacheRegistry.get(dedupedFn);
  if (!requestStore) {
    throw new Error("dedupe: cache not found");
  }
  const { headers } = await Promise.resolve().then(() => _interopRequireWildcard(require("next/headers")));
  const h = await headers();
  return requestStore.delete(h);
}

// src/next/index.ts
var evaluationCache = /* @__PURE__ */ new WeakMap();
function getCachedValuePromise(headers, flagKey, entitiesKey) {
  const map = _optionalChain([evaluationCache, 'access', _2 => _2.get, 'call', _3 => _3(headers), 'optionalAccess', _4 => _4.get, 'call', _5 => _5(flagKey)]);
  if (!map) return void 0;
  return map.get(entitiesKey);
}
function setCachedValuePromise(headers, flagKey, entitiesKey, flagValue) {
  const byHeaders = evaluationCache.get(headers);
  if (!byHeaders) {
    evaluationCache.set(
      headers,
      /* @__PURE__ */ new Map([[flagKey, /* @__PURE__ */ new Map([[entitiesKey, flagValue]])]])
    );
    return;
  }
  const byFlagKey = byHeaders.get(flagKey);
  if (!byFlagKey) {
    byHeaders.set(flagKey, /* @__PURE__ */ new Map([[entitiesKey, flagValue]]));
    return;
  }
  byFlagKey.set(entitiesKey, flagValue);
}
var transformMap = /* @__PURE__ */ new WeakMap();
var headersMap = /* @__PURE__ */ new WeakMap();
var cookiesMap = /* @__PURE__ */ new WeakMap();
var identifyArgsMap = /* @__PURE__ */ new WeakMap();
function transformToHeaders(incomingHeaders) {
  const cached = transformMap.get(incomingHeaders);
  if (cached !== void 0) return cached;
  const headers = new Headers();
  for (const [key, value] of Object.entries(incomingHeaders)) {
    if (Array.isArray(value)) {
      value.forEach((item) => {
        headers.append(key, item);
      });
    } else if (value !== void 0) {
      headers.append(key, value);
    }
  }
  transformMap.set(incomingHeaders, headers);
  return headers;
}
function sealHeaders(headers) {
  const cached = headersMap.get(headers);
  if (cached !== void 0) return cached;
  const sealed = _chunkZFCSUOEScjs.HeadersAdapter.seal(headers);
  headersMap.set(headers, sealed);
  return sealed;
}
function sealCookies(headers) {
  const cached = cookiesMap.get(headers);
  if (cached !== void 0) return cached;
  const sealed = _chunkZFCSUOEScjs.RequestCookiesAdapter.seal(new (0, _cookies.RequestCookies)(headers));
  cookiesMap.set(headers, sealed);
  return sealed;
}
function isIdentifyFunction(identify) {
  return typeof identify === "function";
}
async function getEntities(identify, dedupeCacheKey, readonlyHeaders, readonlyCookies) {
  if (!identify) return void 0;
  if (!isIdentifyFunction(identify)) return identify;
  const args = identifyArgsMap.get(dedupeCacheKey);
  if (args) return identify(...args);
  const nextArgs = [
    { headers: readonlyHeaders, cookies: readonlyCookies }
  ];
  identifyArgsMap.set(dedupeCacheKey, nextArgs);
  return identify(...nextArgs);
}
function getDecide(definition) {
  if (definition.adapter && typeof definition.adapter.decide !== "function") {
    throw new Error(
      `flags: You passed an adapter that does not have a "decide" method for flag "${definition.key}". Did you pass "adapter: exampleAdapter" instead of "adapter: exampleAdapter()"?`
    );
  }
  if (typeof definition.decide !== "function" && typeof _optionalChain([definition, 'access', _6 => _6.adapter, 'optionalAccess', _7 => _7.decide]) !== "function") {
    throw new Error(
      `flags: You passed a flag declaration that does not have a "decide" method for flag "${definition.key}"`
    );
  }
  return function decide(params) {
    if (typeof definition.decide === "function") {
      return definition.decide(params);
    }
    if (typeof _optionalChain([definition, 'access', _8 => _8.adapter, 'optionalAccess', _9 => _9.decide]) === "function") {
      return definition.adapter.decide({ key: definition.key, ...params });
    }
    throw new Error(`flags: No decide function provided for ${definition.key}`);
  };
}
function getIdentify(definition) {
  return function identify(params) {
    if (typeof definition.identify === "function") {
      return definition.identify(params);
    }
    if (typeof _optionalChain([definition, 'access', _10 => _10.adapter, 'optionalAccess', _11 => _11.identify]) === "function") {
      return definition.adapter.identify(params);
    }
    return definition.identify;
  };
}
function getRun(definition, decide) {
  return async function run(options) {
    let readonlyHeaders;
    let readonlyCookies;
    let dedupeCacheKey;
    if (options.request) {
      const headers = transformToHeaders(options.request.headers);
      readonlyHeaders = sealHeaders(headers);
      readonlyCookies = sealCookies(headers);
      dedupeCacheKey = options.request.headers;
    } else {
      const { headers, cookies } = await Promise.resolve().then(() => _interopRequireWildcard(require("next/headers")));
      const [headersStore, cookiesStore] = await Promise.all([
        headers(),
        cookies()
      ]);
      readonlyHeaders = headersStore;
      readonlyCookies = cookiesStore;
      dedupeCacheKey = headersStore;
    }
    const overrides = await getOverrides(
      _optionalChain([readonlyCookies, 'access', _12 => _12.get, 'call', _13 => _13("vercel-flag-overrides"), 'optionalAccess', _14 => _14.value])
    );
    const entities = await getEntities(
      options.identify,
      dedupeCacheKey,
      readonlyHeaders,
      readonlyCookies
    );
    const entitiesKey = _nullishCoalesce(JSON.stringify(entities), () => ( ""));
    const cachedValue = getCachedValuePromise(
      readonlyHeaders,
      definition.key,
      entitiesKey
    );
    if (cachedValue !== void 0) {
      _chunkZFCSUOEScjs.setSpanAttribute.call(void 0, "method", "cached");
      const value = await cachedValue;
      return value;
    }
    if (overrides && overrides[definition.key] !== void 0) {
      _chunkZFCSUOEScjs.setSpanAttribute.call(void 0, "method", "override");
      const decision2 = overrides[definition.key];
      setCachedValuePromise(
        readonlyHeaders,
        definition.key,
        entitiesKey,
        Promise.resolve(decision2)
      );
      _chunkZFCSUOEScjs.internalReportValue.call(void 0, definition.key, decision2, {
        reason: "override"
      });
      return decision2;
    }
    const decisionPromise = (async () => {
      return decide({
        // @ts-expect-error TypeScript will not be able to process `getPrecomputed` when added to `Decide`. It is, however, part of the `Adapter` type
        defaultValue: definition.defaultValue,
        headers: readonlyHeaders,
        cookies: readonlyCookies,
        entities
      });
    })().then(
      (value) => {
        if (value !== void 0) return value;
        if (definition.defaultValue !== void 0)
          return definition.defaultValue;
        throw new Error(
          `flags: Flag "${definition.key}" must have a defaultValue or a decide function that returns a value`
        );
      },
      (error) => {
        if (isInternalNextError(error)) throw error;
        if (definition.defaultValue !== void 0) {
          if (process.env.NODE_ENV === "development") {
            console.info(
              `flags: Flag "${definition.key}" is falling back to its defaultValue`
            );
          } else {
            console.warn(
              `flags: Flag "${definition.key}" is falling back to its defaultValue after catching the following error`,
              error
            );
          }
          return definition.defaultValue;
        }
        console.warn(
          `flags: Flag "${definition.key}" could not be evaluated`
        );
        throw error;
      }
    );
    setCachedValuePromise(
      readonlyHeaders,
      definition.key,
      entitiesKey,
      decisionPromise
    );
    const decision = await decisionPromise;
    if (_optionalChain([definition, 'access', _15 => _15.config, 'optionalAccess', _16 => _16.reportValue]) !== false) {
      _chunkZFCSUOEScjs.reportValue.call(void 0, definition.key, decision);
    }
    return decision;
  };
}
function getOrigin(definition) {
  if (definition.origin) return definition.origin;
  if (typeof _optionalChain([definition, 'access', _17 => _17.adapter, 'optionalAccess', _18 => _18.origin]) === "function")
    return definition.adapter.origin(definition.key);
  return _optionalChain([definition, 'access', _19 => _19.adapter, 'optionalAccess', _20 => _20.origin]);
}
function flag(definition) {
  const decide = getDecide(definition);
  const identify = getIdentify(definition);
  const run = getRun(definition, decide);
  const origin = getOrigin(definition);
  const api = _chunkZFCSUOEScjs.trace.call(void 0, 
    async (...args) => {
      _chunkZFCSUOEScjs.setSpanAttribute.call(void 0, "method", "decided");
      if (typeof args[0] === "string" && Array.isArray(args[1])) {
        const [precomputedCode, precomputedGroup, secret] = args;
        if (precomputedCode && precomputedGroup) {
          _chunkZFCSUOEScjs.setSpanAttribute.call(void 0, "method", "precomputed");
          const value = await getPrecomputed(
            api,
            precomputedGroup,
            precomputedCode,
            secret
          );
          if (value === void 0) return definition.defaultValue;
          return value;
        }
      }
      if (args[0] && typeof args[0] === "object" && "headers" in args[0]) {
        const [request] = args;
        return run({ identify, request });
      }
      return run({ identify, request: void 0 });
    },
    {
      name: "flag",
      isVerboseTrace: false,
      attributes: { key: definition.key }
    }
  );
  api.key = definition.key;
  api.defaultValue = definition.defaultValue;
  api.origin = origin;
  api.options = _chunkWECCP5XGcjs.normalizeOptions.call(void 0, definition.options);
  api.description = definition.description;
  api.identify = identify ? _chunkZFCSUOEScjs.trace.call(void 0, identify, {
    isVerboseTrace: false,
    name: "identify",
    attributes: { key: definition.key }
  }) : identify;
  api.decide = _chunkZFCSUOEScjs.trace.call(void 0, decide, {
    isVerboseTrace: false,
    name: "decide",
    attributes: { key: definition.key }
  });
  api.run = _chunkZFCSUOEScjs.trace.call(void 0, run, {
    isVerboseTrace: false,
    name: "run",
    attributes: { key: definition.key }
  });
  return api;
}
function getProviderData(flags) {
  const definitions = Object.values(flags).filter((i) => !Array.isArray(i)).reduce((acc, d) => {
    acc[d.key] = {
      options: d.options,
      origin: d.origin,
      description: d.description,
      defaultValue: d.defaultValue,
      declaredInCode: true
    };
    return acc;
  }, {});
  return { definitions, hints: [] };
}













exports.clearDedupeCacheForCurrentRequest = clearDedupeCacheForCurrentRequest; exports.combine = combine; exports.createFlagsDiscoveryEndpoint = createFlagsDiscoveryEndpoint; exports.dedupe = dedupe; exports.deserialize = deserialize2; exports.evaluate = evaluate; exports.flag = flag; exports.generatePermutations = generatePermutations; exports.getPrecomputed = getPrecomputed; exports.getProviderData = getProviderData; exports.precompute = precompute; exports.serialize = serialize2;
